README
My own fork based on jacob1's mod.

I aim to make realistic stuff and would love if physics was not this bad on tpt :(, but oh well, I'll have to make it do with what I've got!











Jacob1's Mod, tpt's longest lasting and most popular mod! My overall goal 
isn't to add a bunch of elements you can forget about, it's to create a 
more stable tpt version without many bugs and lots of new features not fit for 
the officiel version; like multiple tabs, an updated save preview and moving 
solids. It it based off the older codebase of tpt version 83, back when things 
were in C and simpler; this allows me to do things how I want without the few 
small but unfixable bugs tpt++ has.

The mod updates whenever I feel like it. Don't ignore the update notifications 
like you do in other mods though, this mod actually has it's own in game 
updater, hosted at https://starcatcher.us/TPT. Typically there is an update,
no matter how small, once every 1-2 months, although that is constantly 
changing as my schedule and interest in working on the mod changes. All the 
changes made to tpt are quickly backported into the mod, making it much more up 
to date and quicker to test new official features.

When you first start up the mod, it might look a lot different (or the same, 
if you remember what TPT used to look like before the rewrite). Everything 
functions exactly the same though. There are a few areas you may want to look 
at to see the new features. The main one is the sub menus in the favorites menu. 
The favorites menu is the first feature added to the mod. Although it is now in 
the official version, the submenus are unique to the mod. You can find many 
options here, all in element button format (i'll make an actual options ui 
someday ....). Elsewhere, if you click the 'T' quickoption in the top right you 
can access the 'tabs' feature. Other changes can be seen in the redone console 
or in the save preview. An in detail explanation of these features and others 
are provided below.




------------------------------------------------------------------------------
------------------------------------Credits-----------------------------------
------------------------------------------------------------------------------
Special Thanks
mniip - hosted the original update server
iam4722202468 - made the bug font icon / good troll friend
cracker64 - made TPTMP, and hosts the update and TPTMP servers


Things in my mod suggested by other people:
Atrayin - Powered Portals, RAZR
OmegaSupreme - Animated LCRY
Videogamer555 - New heat displays
disturbed666 (aka grandmaster) - A way to turn PHOT into ELEC (but I used PINV instead of TESC)
Dynotec - new statistics
therocketeer - smudge tool
cip - INWR with tmp of 1 doesn't conduct to SWCH
BoredInSchool - fusion
tommig - VOID ctypes
Galacticruler - SING can emit gravity instead of pressure (set tmp2 to 1, or 2 for both)
Joeboy25 - AMTR ctypes
baizuo - Multiple element highlights
The-Fall - GEL and SPNG can absorb water from more elements (ex. PSTE + SPNG -> CLST)
GIGATeun - GRVI (gravitons)
minecraft-physics - DRAY (duplicator ray), CRMC
G-LinuxorU - CRMC additions


2014 April fools joke element suggestions:
therocketeer - DKLQ (dinosaur knee liquid)
CatAnimation - BLOD/COCO/COCJ/LOLZ2
Darthanihlus - POOP
cracker64 - CMD
bowserinator - DMD2/BDMD (breakable diamond)
Kikinicij - BATT (ban all the things)
Nakriollag - DIMT




------------------------------------------------------------------------------
---------------------------------Explanations---------------------------------
------------------------------------------------------------------------------
Favorites Menu / Submenus
The favorites menu has been the center of most things in the mod since the 
beginning. To "favorite" an element, ctrl+shift click it from any menu, and 
it will get a yellow heart by it. To unfavorite it, ctrl+shift+right click 
it instead. Any recently used elements also will appear in the menu.
If you click the MORE button, a new menu will appear with more options. 
HUD's are customizable using the HUD2 button. The FIND button colors whatever 
element you have selected red, and dims everything else (now in official TPT, 
with the ctrl+f shortcut). There's also a way to double or triple highlight 
things in green and blue at the same time, if you click the button more times. 
FND2 changes the way find is displayed for layered particles. The INFO button 
displays statistics about how long and how many times you have played powder 
toy, your average FPS, and info about the current save. REAL turns on realistic 
heat mode by savask. DATE changes date and time settings, used in the HUD and on 
save comments. In addition to left clicking, right clicking it will toggle 
displaying time on comments older than a day. The rest (SPIN, HEAT, LUA) are 
explained later in the readme.

Moving Solids
No lag, and it actually works. They are found in the special menu, and you 
can have a maximum of 256 at once. They are drawn the same shape as your 
brush, no matter what shape or size it is. They are very bouncy, and fall 
with gravity. Pressure over 10.0 will destroy them, and if the center 
particle is destroyed, it will fall apart, making really tiny bouncing 
pixels. Go into the FAV2 menu and click SPIN to make them able to rotate, now 
enabled by default. Recent version 29 additions allow them to properly move 
around in loop edge mode and to bounce off walls. Version 30 rewrites them 
internally, while paused any moving solids you draw will all be attached as 
one ball (even if they aren't attached physically). This allows you to create  
unique shapes.

Animated LCRY
Found in the powered elements menu. In the decoration editor, 
press left or right to change frames, and paint each frame normally. Press 
Del to delete a frame or press Ctrl+right when going to a new frame to copy 
the old one over to the new one. PSCN activates it, and the speed that it 
changes depends on its temperature. NSCN deactivates it. METL pauses/unpauses
it. If you set the tmp2 of the PSCN or NSCN that (de)activates it, you can 
set which frame ANIM starts or freezes at. Also, you can use the tmp of the 
PSCN to set the delay for the first frame. You can have a maximum of 25 
frames, unless you use the lua command tpt.maxframes(num), where num is the 
new maximum, between 1 and 256. Using that command will reset all existing 
ANIM. If you delete a lot of it at once, the game will freeze for a while to 
free all the allocated memory, just wait and it will keep working.

New Heat Displays
Found in the FAV2 menu. Click the HEAT button once to go to automatic display 
mode, where the hottest temp is always pink, the lowest is dark blue, and 
everything else in between is based off of that. Click it again to go into 
manual heat display. The min and max temps will stop changing when you go 
into this, so it will help if automatic mode flickers because of something 
like CLNE(FIRE) constantly changing the max temp. If you right click the 
button, it will ask you to enter the new maximum and minimum temp. Enter the 
temperatures in Celsius. It will change the display mode into manual 
automatically after this.

Powered Elements
All powered elements are now activated instantly. When you start the mod or 
clear the sim, a setting is set which causes all powered elements to be 
instantly activated. Saves loaded from official TPT don't have this set, so 
will maintain compatibility. Elements from the mod are always activated 
instantly regardless of this setting. Also, if you want to make a powered 
element in Lua it's as simple as setting PROP_POWERED, it will handle the 
life values for you (you will have to do the graphics though).

Lua Code
Put a Lua script in luacode.txt and click the LUA button in the FAV2 menu. It 
will read the script and then run it once the following frame. If you need 
something to run every frame, use tpt.register_step. Once you do this the 
script becomes "embeded" in every save and stamp you create. When you load the 
save, it will write the code back out to newluacode.txt and ask if you want 
to run it. This is just so that you can see what you are running and look for 
malicious code before it is executed. Clearing the simulation will also make 
it forget about any code, although there may still be side effects from the 
script.

Custom HUDs
Inside the HUD2 menu you can click on HUD2 to make a customized HUD. By 
default the mod has a custom HUD on that fixes many issues of the original 
HUD (but these fixes are in TPT now). You can also customize the original HUD 
too (toggled with HUD button), and both debug mode HUDs. Theoretically, you 
could have 4 HUDs saved at once, and the HUD2 menu will edit whichever is 
enabled at the time. After clicking HUD2, there are more submenus. RSET will 
reset all of the HUDs back to their default values. CORD lets you toggle 
options related to the second row in the debug HUD. INFO lets you change 
options related to the uper right line. UI lets you change the things 
displayed in the upper left. Most buttons in these submenus are toggles, but 
anything ending in # brings up an input box and allows you to set how many 
decimal places things like pressure and temperature are displayed to.

Tabs
Tabs are shown in the 'T' quickoption on the upper right. clicking it toggles 
showing tabs, you can also hold control to view the tab list. Tabs are 
basically alternate simulations, which save everything including pressure and 
current save info. A few really advanced settings may not be saved, as behind 
the scenes it is simply saving a stamp. They aren't simulated when you aren't 
looking at them but still have uses. To save the current tab to disk, press 
ctrl+s. You can then use the reload button from this point, even without an 
account or looking at a save from the save server. You could just stop here 
with the tab feature, and only use a single tab for reloading, but there is a 
lot more. To create a new tab, press ctrl+n or click the '+' icon in the tab 
menu. Your previous tab will be saved. When hovering over tabs it will 
display a thumbnail preview and the name of the save so you can quickly 
identify them. Right clicking on a tab will delete it.
Two obscure features to note: If you are on the main tpt screen and you open a
save from the tpt website, it will open it in a new tab instead of launching 
a new instance of tpt. Also, if your game crashes for any reason, it will 
save your current tab and restart TPT automatically with all of your previous 
tabs intact.

Console
The console has been redone to fix many issues the original one had. Text 
will no longer overlap, it properly wraps text on both the input and output 
sides, allowing you to see overly long commands. You can drag the line in the 
middle of the console to adjust how much room either side gets. The console
also now scrolls, allowing you to see the previous 20 commands and command
output. When typing a command the input will also wrap around, fixing the
arbitrarily short limit of the console in vanilla TPT. In addition, console
output is saved to powder.pref so you can see what previous commands returned.
Syntax highlighting from vanilla TPT has not been added yet.

Save Preview
This may not look a lot different than the default TPT, because tpt++ 
overhauled the save preview itself when it was redone in c++. It used to be 
cutting edge though! My mod doesn't have avatars or a scrollbar yet, but it 
does have other cool features. Comment scrolling works with the scrollwheel on 
your mouse, or the '[' and ']' keys. There are no 'pages' like in tpt++, it's 
just a continuous stream of comments. It will attempt to display the page 
number tpt would have on bottom, but at the moment it can't properly display 
the maximum number of pages due to API limitations. To the right of each name 
is the date and time each comment was posted. You can edit how this is 
displayed with the DATE option in the fav2 menu. If you click it, it will 
toggle showing User ID's instead. Clicking the 'Submit' button without anything 
in the textbox will reload the comments. The comment box turns red when the 
comment is too long to be submitted (the server currently has a 500 character 
limit). Left clicking a username will append their nick to the end of the 
textbox to make it easier to reply to their comment without having to type out 
the username. Shift-clicking their username will open a search for all of their 
public saves. Ctrl-clicking their username will open the profile UI with their 
bio and avatar.




------------------------------------------------------------------------------
------------------------------------History-----------------------------------
------------------------------------------------------------------------------
The original point of this mod was the favorite elements menu back in version 
1. It was intended to do everything my badly written lua script did, because 
several users such as @MasterMind555 encouraged me to start developing tpt. 
It was very ambitious, I barely knew how to code and claimed version 2 would 
have moving solids, something everyone said was impossible. Version 1.0 
was released on October 18th, 2011, with the Version 2.0 moving solids update 
shortly after. I added many random suggestions and almost all the current 
elements within the few months after the original release. Over time, I have 
fixed a lot of the original things (moving solids back then were hollow and 
had bad collisions) as I figured out what I was doing.

At some point, I decided to add @BoredInSchool's fusion suggestion into my 
mod. Everyone loved it, and I decided to try and get some of my code into the 
official game. After two pull requests filled with fusion, lots of bug fixes, 
and other small features, I was unexpectedly made a moderator. A few months 
later I was given access to the repo to help fix all the bugs in the c++ 
rewrite. Three times before I have stopped work on this for periods of a month 
or more, but I always come back to it and keep it up to date as new tpt 
versions are released. Originally, I had planned to port my mod into the c++ 
rewrite in hopes it would be merged into official. Eventually I realized that 
was never going to happen, and in fact I didn't want this in official. I like 
being able to do whatever I want without having others tell me what needs to 
be done and what can't be put in. I went into full backporting mode, putting 
all the missing features from the c++ rewrite into my mod so that you 
wouldn't miss any features when using this. Even though i'm really busy with 
college, I am still doing this today, and now plan to rewrite a few parts 
into c++ myself so that I can add more features.

Sometime in the middle, I thought the idea of having a special update server 
would be cool. @mniip offered to host it for me, with his new server. I 
quickly put something together that even had in game changelogs, and he did 
the server side script. It worked great, this helped keep people interested 
in the mod, because they didn't need to redownload every new version, and 
eventually get tired of it. Now, a lot of people actually use this as an 
entire tpt replacement (including me).




------------------------------------------------------------------------------
-------------------------------------Bugs-------------------------------------
------------------------------------------------------------------------------
None (at least that I know of)




------------------------------------------------------------------------------
-------------------------------------TODO-------------------------------------
------------------------------------------------------------------------------
TODO: '>' means currently being worked on/added next
>Use new tpt++ methods to fetch saves and not over 7 year old ones.
>Rewrite save browser and save preview UI to new interface
Support alternate save servers
Support multiple logins / "remember me" setting
make open ui more colorful like tpt++, also fix cut off description in save uploading ui
Ability to follow users using jacksonmj's website
>Add things from tpt++ that look cool, and any new simulation changes it gets

Possible future plans:
customizable settings menu completely, ability to add and read options. (also HUD option settings)
CMND - command element to run old commands
finish what was started and add ability for multiple Simulation*s to run at once?
Things never backported: lua panic, regex removal, old/lua syntax highlighting (https://github.com/simtr/The-Powder-Toy/commit/ecf1f0e3917386f087dbe569862bd6486a4b366a)
ability to hide certain saves (based on things like number of votes, unpublished / published, username, etc.)


CHANGELOG

Version 1.0: http://min.us/lbuOtsPdEMibD2
Original release, with most of the stuff from my Lua Mod

Version 2.0: http://min.us/lzFWVBYhl5Xfi
Now with MOVING SOLIDS!

Version 2.1: http://min.us/lbvyL8rkNJlcni
Latest source + source now on github

Version 3.0: http://min.us/lbjHzEoxmXAFzP
Added Animated LCRY. Moving solids and ANIM save when you make a stamp or 
save.

Version 4.0: http://min.us/l2FzyIz4AYBS9
Moving solids can rotate now, disabled by default, 
go into the Fav2 menu and click SPIN to allow them to rotate. 
Also added Indestructible insulator and One time wire, 
and did some fixes to FIND and BALL.

Version 5.0: http://min.us/lbh6DDTJzdNFU3
Added automatic and manual heat displays. In automatic heat display, 
the highest temp is pink and the lowest is dark blue. In Manual, 
you can decide which temp is displayed as the hottest color and which is 
the lowest, and everything in between will be based off that. 
I also added powdered portals and virus/cure, and updated to the latest 
source to get rid of the update box.

Version 5.1: http://min.us/lDCKmb9BdFazn
Updated to latest version, added activator, which can be used in creations 
to prevent people from sparking buttons until specific times. Also, re-added 
in buttons to switch to the different modes, except blob and heat gradient.

Version 6.0: http://min.us/lUgHNsGndzdbh
Added powered invisible, blob and heat gradient display, portal effects, and 
wifi/portal lines. Also added the old lua console commands: sound(must be 
enable with a shortcut), load, bubble, reset_temp, and reset_pressure(works 
with the same parameters reset_velocity does). There are three new lua 
commands: get_pressure, get_gravity, and maxframes(sets the number of frames 
ANIM can have when created, 1-256). Also, render_ui uses the old cmode 
pictures in color.Note: the button keys to change display modes, blob & heat 
gradient display, portal effects (Simon forgot the wifi/portal lines), and 
new icons all got into the official version. Not much of this code was 
actually written by me, a lot of it was taken from github after Simon deleted 
it. He said he might add in more stuff that I wrote later.

Version 7.0: http://min.us/lK604TH6bFNV4
Added readme. Faster drawing and deleting of elements. Added RAZR, an element
heavier than all other elements. Added new button in FAV2 menu that can make
saves compatible with the latest beta or release version and able to be
published. It makes sure that you don't add in any new elements not in those
versions. PINV+PHOT = ELEC. SPNG drags along other particles with it, even 
solids that normally can't move.

Version 8.0: http://min.us/muN92vQxC
Wall names display in the hud. New lua functions: get_wall and create_wall. A 
New Wall: ERASEALL. It can erase walls, signs, and particles all at once. 
Most powered elements drawn in this version are now activated instantly. Set 
the tmp to zero to keep the old way of activating. New properties that can be 
set using the lua console: CLONE (0x40000), BREAKABLECLONE (0x80000), 
POWERED (0x100000), and INDESTRUCTIBLE (0x20000). Use the first 3 together to 
make different kinds of clone using elements. The last one makes things 
invincible to every element, and even prevents it from being destroyed in 
reactions. You can make things like indestructible WATR or METL. Lua code can 
now be saved. Put it in luacode.txt, and it will be saved and run in 
stamps/saves. Press up in the console and then enter to load an example.

Version 9: http://min.us/muN92vQxC
Fix loading some saves and fixed crash with HUD. Removed update button 
permanently and updated to newer source. Also, decoration color is now drawn 
on stamps in the stamp browser. 2 new quickoptions: Newtonian gravity and 
ambient heat. Custom HUDs. To change one, click the HUD2 button in the second 
favorites menu and different options will come up. You can change all 4 of 
the original HUD's and it will save and be loaded the next time you start. 
Also added a few more options not in any normal HUD.

Version 10
Added display mode toggles, press Ctrl+ a number key besides 4 and 8 to 
toggle the main feature of that display temporarily. Added more statistics. 
Fixed some moving solid bugs and made rotation save. Updated to newest 
source, which includes the new save format, but you can choose to save in the 
old format instead. You can delete corrupted stamps now. Fix three powered 
electronic related bugs, all electronics should work like they're supposed 
to now. Added a decoration smudge (blend) tool

Version 11 - 2/21/12
Added PROP2 tool, you can draw properties on with the normal brush now. 
Improved decoration drawing speed and added a hex value display and floodfill 
to it. Changed menus of some elements. Added a new test icon for the favorites 
menu. Added an update check, if there is a new version, you can use the link 
in the message of the day to download a zip file containing the next version 
of this mod. Added COND, only conducts to other COND with same tmp within a 
tmp2 radius. Added PWHT, flood fill heats the element above it to it's temp. 
Use PROP on it to make it create other properties. Added rechargeable battery, 
set the tmp of battery to the maximum charge it can hold, METL to charge, 
PSCN/NSCN to turn on/off. Some other stuff not listed here.

Version 12 - 3/2/12
Added tabs. Hit the tab button in the quickoptions menu or press Ctrl to show 
a menu of all the tabs you have open. You can switch between them, or open a 
new one (what you are doing is copied over). Tabs are saved and not simulated 
when you are not using them. Press shift to save pressure in stamps and saves, 
RCTRL to keep it when transforming stamps. Added fusion. see thread 12901 for 
more info (http://powdertoy.co.uk/Discussions/Thread/View.html?Thread=12901). 
Added realistic mode, not by me and not too noticeable. Autosave button, 
disabled by default. If Powder Toy unexpectedly closes, the next time you 

includes/defines.h
/**
 * Powder Toy - Main source
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#ifndef DEFINE_H
#define DEFINE_H

#ifdef DEBUG
#include <iostream>
#endif

#ifdef WIN
#define PATH_SEP "\\"
#else
#define PATH_SEP "/"
#endif

//VersionInfoStart
#define SAVE_VERSION 95
#define MINOR_VERSION 0
#define BUILD_NUM 345
//VersionInfoEnd
#define FAKE_SAVE_VERSION 95
#define FAKE_MINOR_VER 0

#define MOD_VERSION 44
#define MOD_MINOR_VERSION 2
#define MOD_SAVE_VERSION 25 //This is not the version number of my mod anymore, it's only changed when I change the saving code
#define MOD_BUILD_VERSION 120 //For update checks

#define MOBILE_MAJOR 1
#define MOBILE_MINOR 9
#define MOBILE_BUILD 115

#define IDENT_VERSION "G" //Change this if you're not Simon! It should be a single letter

#define MTOS_EXPAND(str) #str
#define MTOS(str) MTOS_EXPAND(str)

#define SCHEME "https://"
#define STATICSCHEME "https://"
#define UPDATESCHEME "https://"
#define ENFORCE_HTTPS

#ifndef SERVER
#define SERVER "powdertoy.co.uk"
#endif
#ifndef STATICSERVER
#define STATICSERVER "static.powdertoy.co.uk"
#endif
#ifndef UPDATESERVER
#define UPDATESERVER "no update"
#endif

#define LOCAL_SAVE_DIR "Saves"

#define THUMB_CACHE_SIZE 256

#ifndef M_PI
#define M_PI 3.14159265f
#endif
#ifndef M_GRAV
#define M_GRAV 6.67300e-1
#endif

#define TIMEOUT 100

#ifdef RENDERER
#define MENUSIZE 0
#define BARSIZE 0
#else
#define MENUSIZE 40
#ifdef TOUCHUI
#define BARSIZE 30
#else
#define BARSIZE 17
#endif
#endif
#define XRES	612
#define YRES	384
#define VIDXRES (XRES+BARSIZE)
#define VIDYRES (YRES+MENUSIZE)
#define NPART XRES*YRES
const int menuStartPosition = XRES+BARSIZE-17;
const int menuIconWidth = 17;

#define XCNTR   XRES/2
#define YCNTR   YRES/2

#define MAX_DISTANCE sqrt(pow(XRES, 2.0f)+pow(YRES, 2.0f))

#define GRAV_DIFF

#define TAG_MAX 256

#define CELL    4
#define ISTP    (CELL/2)
#define CFDS	(4.0f/CELL)
#define SIM_MAXVELOCITY 1e4f

#define AIR_TSTEPP 0.3f
#define AIR_TSTEPV 0.4f
#define AIR_VADV 0.3f
#define AIR_VLOSS 0.999f
#define AIR_PLOSS 0.9999f

#define GRID_X 5
#define GRID_Y 4
#define GRID_P 3
#define GRID_S 6
#define GRID_Z 3

#define CATALOGUE_X 4
#define CATALOGUE_Y 3
#define CATALOGUE_S 6
#define CATALOGUE_Z 3

#define STAMP_MAX 5000

//#define SAVE_OPS
//#define REALISTIC

#define SURF_RANGE     10
#define NORMAL_MIN_EST 3
#define NORMAL_INTERP  20
#define NORMAL_FRAC    16

#define REFRACT        0x80000000

/* heavy flint glass, for awesome refraction/dispersion
   this way you can make roof prisms easily */
#define GLASS_IOR      1.9f
#define GLASS_DISP     0.07f

#ifdef _MSC_VER
#pragma warning(disable: 4100) // unreferenced formal parameter
#endif

#define SDEUT

#define DEBUG_PARTS				0x0001
#define DEBUG_ELEMENTPOPULATION	0x0002
#define DEBUG_DRAWTOOL			0x0004
#define DEBUG_PARTICLE_UPDATES	0x0008

extern bool firstRun;
extern bool showLargeScreenDialog;
extern int screenWidth;
extern int screenHeight;
extern float FPSB2;
extern int elapsedTime;

extern int NUM_PARTS;

extern bool legacy_enable;
extern bool aheat_enable;
extern bool decorations_enable;
extern int active_menu;
extern int last_active_menu;
extern int last_fav_menu;
extern bool hud_enable;
extern bool pretty_powder;
extern bool drawgrav_enable;
extern int limitFPS;
extern bool water_equal_test;
extern int finding;
extern int foundParticles;
extern int highesttemp;
extern int lowesttemp;
extern int heatmode;
extern int secret_els;
extern int tab_num;
extern int num_tabs;
extern bool show_tabs;
#ifdef __cplusplus
class Tool;
extern Tool* activeTools[3];
#endif
extern float toolStrength;
extern int autosave;
extern int realistic;
extern bool explUnlocked;
extern int old_menu;
extern int decobox_hidden;

extern int drawinfo;
extern int currentTime;
extern int totaltime;
extern int totalafktime;
extern int afktime;
extern int frames;
extern double totalfps;
extern int prevafktime;
extern int timesplayed;

extern int deco_disablestuff;

extern bool doUpdates;

extern int debug_flags;
#define DEBUG_PERF_FRAMECOUNT 256
extern int debug_perf_istart;
extern int debug_perf_iend;
extern long debug_perf_frametime[DEBUG_PERF_FRAMECOUNT];
extern long debug_perf_partitime[DEBUG_PERF_FRAMECOUNT];
extern long debug_perf_time;

extern int active_menu;

extern bool sys_pause;
extern int framerender;

#include "graphics/Pixel.h"
struct stamp
{
	char name[11];
	pixel *thumb;
	int thumb_w, thumb_h, dodelete;
};
typedef struct stamp stamp;

extern bool console_mode;
extern bool REPLACE_MODE;
extern bool SPECIFIC_DELETE;
extern int GRID_MODE;
extern int DEBUG_MODE;
extern stamp stamps[STAMP_MAX];
extern int stamp_count;

extern int ptsaveOpenID;
extern int saveURIOpen;
extern int do_open;
extern pixel *vid_buf;

extern int scrollSpeed;
extern float scrollDeceleration;

extern unsigned short last_major, last_minor, update_flag, last_build;

extern char http_proxy_string[256];

//Functions in main.c
void thumb_cache_inval(char *id);
void thumb_cache_add(char *id, void *thumb, int size);
bool thumb_cache_find(char *id, void **thumb, int *size);
void clear_sim();
void NewSim();
char* stamp_save(int x, int y, int w, int h, bool includePressure);
void tab_save(int num);
#ifdef __cplusplus
class Save;
Save *stamp_load(int i, int reorder);
#endif
int tab_load(int tabNum, bool del = false, bool showException = true);
void stamp_init();
void del_stamp(int d);
void ctrlzSnapshot();

extern bool sendNewEvents;
extern bool openConsole;
extern bool openSign;
extern bool openProp;
#ifdef __cplusplus
class PowderToy;
extern PowderToy *the_game;
#endif
int main_loop_temp(int b, int bq, int sdl_key, int scan, int x, int y, bool shift, bool ctrl, bool alt);
void main_end_hack();
#endif
